tests/unit/sidecars/secrets_fetcher_tests.py

```diff
@@ -0,0 +1,220 @@
+vault.url = https://vault.example.com:8200/
+vault.role = my-server-role
+vault.auth_type = aws
+vault.mount_point = aws-ec2
+
+output.path = /var/local/secrets.json
+output.owner = {getpass.getuser()}
+output.group = {getpass.getuser()}
+output.mode = 0400
+
+secrets =
+    secret/one,
+    secret/two,
+    secret/three,
+
+callback = scripts/my-transformer  # optional
+""".strip()
+
+
+@dataclasses.dataclass
+class FakeVaultClient(secrets_fetcher.VaultClient):
+    token_expiration: datetime.datetime
+    token: str = "token"
+
+    # @typing.override  # TODO: Added in version 3.12.
+    def get_secret(self, secret_name: str) -> tuple[typing.Any, datetime.datetime]:
+        return secret_name.upper(), self.token_expiration + datetime.timedelta(seconds=30)
+
+
+@dataclasses.dataclass
+class FakeBadVaultClient(secrets_fetcher.VaultClient):
+    token_expiration: datetime.datetime
+    token: str = "token-bad"
+
+    # @typing.override  # TODO: Added in version 3.12.
+    def get_secret(self, secret_name: str) -> tuple[typing.Any, datetime.datetime]:
+        """Return a secret value that is not JSON serializable."""
+        s = self.token_expiration
+        return s, self.token_expiration + datetime.timedelta(seconds=30)
+
+
+class Tests(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        spec = {
+            "vault": {
+                "url": config.DefaultFromEnv(config.String, "BASEPLATE_DEFAULT_VAULT_URL"),
+                "role": config.String,
+                "auth_type": config.Optional(
+                    config.OneOf(**secrets_fetcher.VaultClientFactory.auth_types()),
+                    default=secrets_fetcher.VaultClientFactory.auth_types()["aws"],
+                ),
+                "mount_point": config.DefaultFromEnv(
+                    config.String, "BASEPLATE_VAULT_MOUNT_POINT", fallback="aws-ec2"
+                ),
+            },
+            "output": {
+                "path": config.Optional(config.String, default="/var/local/secrets.json"),
+                "owner": config.Optional(config.UnixUser, default=0),
+                "group": config.Optional(config.UnixGroup, default=0),
+                "mode": config.Optional(config.Integer(base=8), default=0o400),  # type: ignore
+            },
+            "secrets": config.Optional(config.TupleOf(config.String), default=[]),
+            "callback": config.Optional(config.String),
+        }
+
+        parser = configparser.RawConfigParser()
+        with io.StringIO(configini) as f:
+            parser.read_file(f)
+        fetcher_config = dict(parser.items("secret-fetcher"))
+
+        cls.cfg = config.parse_config(fetcher_config, spec)
+
+    def setUp(self):
+        self.setUpPyfakefs()
+        self.fake_fs().create_file("/var/local/secrets.json", contents="initial contents")
+
+        cfg = self.cfg
+        now = datetime.datetime.now(UTC)
+        with unittest.mock.patch("baseplate.sidecars.secrets_fetcher.VaultClientFactory") as mock:
```

Comment:
Can you set `autospec=True` here? IMO this should basically always be used when mocking like this in order to avoid replacing the class with a MagicMock that doesn't enforce you call it like the original class it's replacing.

Reply:
done.